### PR TITLE
Add clippy to `toolstate.toml`

### PR DESCRIPTION
**Pull request by oli-obk**
r? @alexcrichton 

cc @Manishearth 

I have no idea how to get clippy working... it needs proc macros, and I think I did everything right (I just did what the cargo step is doing), but it's not working:

```
error: libproc_macro-6210e4b46662ec28.so: cannot open shared object file: No such file or directory
  --> src/tools/clippy/clippy_lints/src/lib.rs:47:1
   |
47 | extern crate serde_derive;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: libproc_macro-6210e4b46662ec28.so: cannot open shared object file: No such file or directory
  --> src/tools/clippy/clippy_lints/src/lib.rs:47:1
   |
47 | extern crate serde_derive;
   | ^
```

It's especially weird since it used to work

Anyway. Fixing it can be left for a future PR, this one adds it to CI, but marks it as "broken"

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @rust-lang/dev-tools, any thoughts about this? The tl;dr; here is that you should be able to just change a line in `src/tools/toolstate.toml` to say "don't build clippy on this PR", so that way it's in theory easy for contributors to not have to go through the whole submodule process to land a PR

File: src/bootstrap/check.rs [hunk 1]
```diff
@@ -348,6 +348,50 @@ impl Step for Miri {
...
+        let host = self.host;
+        let compiler = builder.compiler(1, host);
+
+        let _clippy = builder.ensure(tool::Clippy { compiler, target: self.host });
+        let mut cargo = builder.cargo(compiler, Mode::Tool, host, "test");
+        cargo.arg("--manifest-path").arg(build.src.join("src/tools/clippy/Cargo.toml"));
+
+        // Don't build tests dynamically, just a pain to work with
+        cargo.env("RUSTC_NO_PREFER_DYNAMIC", "1");
+        // miri tests need to know about the stage sysroot
```
Comment (Manishearth) [inline src/bootstrap/check.rs]: s/miri/clippy?

do we need to still know the sysroot?

**Comment by Manishearth** [comment]
I'm +1 on this. We should probably encourage that folks try to get a working patch to fix it at least (even if they then hand it off to someone else to coordinate landing). But the general tooling sounds good and easy to use.

**Comment by nrc** [comment]
I'm +1 on adding Clippy to toolstate.toml. I don't think we should enable testing by default at this stage - I'm not confident enough of the workflow here just yet (I'm in favour of taking this super-slowly because I worry about causing issues for compiler devs or of tools getting broken and disabled but not fixed, and I think it is worth getting the workflow right for all tools, rather than rushing to push Clippy forward).

**Comment by Manishearth** [comment]
(Since this confused me, nrc clarified that he's only referring to postponing _testing_, not _building_ clippy.)

**Comment by oli-obk** [comment]
I would prefer to allow changing the toolstate to testing, if only so we get pinged when it happens and the state is downgraded to building. It is not possible to change the toolstate to broken if it is still building. So there won't be any "overreaction" on the Dev side.

That said, I can make the testing pass not run by default, which will mean that even if the state is changed to testing, it'll require explicitly running `./x.py test src/tools/clippy` in order for the testing state to act differently from the compiling state

**Comment by oli-obk** [comment]
I disabled clippy testing without passing a full path

File: src/bootstrap/check.rs (hunk 1 above)
Comment (oli-obk) [reply src/bootstrap/check.rs]: fixed

Yea we need it (https://github.com/rust-lang-nursery/rust-clippy/blob/cargo_clippy/src/driver.rs#L146)

**Comment by oli-obk** [comment]
* Travis likes it now
* clippy is set to "Broken" until the proc-macro thing has been fixed (any help is appreciated)
* setting clippy to "Testing" is just the same as setting it to "Building" except when explicitly calling `./x.py test src/tools/clippy`

**Comment by alexcrichton** [comment]
Just to ensure that we're continuing to take this slowly, can this continue to disable clippy by default?

**Comment by oli-obk** [comment]
Done

**Pull request merged**